### PR TITLE
Self-publishing components

### LINKED ISSUE #1319: Have an easy way of Components to publish themselves in the Injector
Currently there is no easy way for a component to publish itself into the `Injector` hierarchy.

## Proposal

* Add a new List property to `@Component` called `publishAs`. If supplied:
* Create a child app injector that publishes the component to child components inside its view (i.e. those that live in the shadow DOM)

## Example

```dart
import 'package:angular2/angular2.dart';

abstract class Container {}

@Component()
@View(
  directives: const [Parent],
  template: '<parent></parent>'
)
class MyApp {
}

@Component(
  publishAs: const [Parent, Container] // <== IMPLEMENT THIS
)
@View(
  template: '<child></child>',
  directives: [Child]
)
class Parent implements Container {
}

@Component()
@View(...)
class Child {
  Child(Container container) {
    // Can't get a hold of Container
  }
}

```

### PR DESCRIPTION
**Pull request by yjbanov**
This fixes #1319

### CODE REVIEW

File: modules/examples/pubspec.yaml [hunk 1]
```diff
@@ -23,9 +23,9 @@ transformers:
         - web/src/hello_world/index.dart
         - web/src/todo/index.dart
 - $dart2js:
-    minify: false
+    minify: true
```
Comment (vicb) [inline modules/examples/pubspec.yaml]: Can we keep it false on examples ?  
_(It could help debugging issues & size should not be sensitive here ?)_

Comment (yjbanov) [inline modules/examples/pubspec.yaml]: This is currently our only code size benchmark so we need it to be `true` in the repo. If you are debugging an issue it should be easy enough to flip it back to `false`.

File: modules/angular2/src/core/compiler/element_injector.js [hunk 2]
```diff
@@ -568,7 +569,13 @@ export class ElementInjector extends TreeNode {
     this._constructionCounter = 0;
   }
 
-  instantiateDirectives(lightDomAppInjector:Injector, host:ElementInjector, shadowDomAppInjector:Injector, preBuiltObjects:PreBuiltObjects) {
+  instantiateDirectives(lightDomAppInjector:Injector, host:ElementInjector,
+      preBuiltObjects:PreBuiltObjects, componentDirective:DirectiveBinding) {
```
Comment (yjbanov) [inline modules/angular2/src/core/compiler/element_injector.js]: remove `componentDirective`; it's already in `this._proto._binding0`

File: modules/angular2/src/core/compiler/element_injector.js [hunk 3]
```diff
@@ -578,6 +585,18 @@ export class ElementInjector extends TreeNode {
 
     var p = this._proto;
     if (isPresent(p._keyId0)) this._getDirectiveByKeyId(p._keyId0);
+    if (hasComponent) {
```
Comment (yjbanov) [inline modules/angular2/src/core/compiler/element_injector.js]: move into `_createShadowDomAppInjector` and only create on injecvtor

File: modules/angular2/test/core/compiler/element_injector_spec.js [hunk 4]
```diff
@@ -576,6 +613,31 @@ export function main() {
         inj.clearDirectives();
         expect(destroy.onDestroyCounter).toBe(1);
       });
+
+      it("should publish component to its children via app injector when requested", function() {
```
Comment (yjbanov) [inline modules/angular2/test/core/compiler/element_injector_spec.js]: also test publishing from dynamic component to dynamically instantiated children

File: modules/angular2/src/core/compiler/element_injector.js (hunk 2 above)
Comment (yjbanov) [inline modules/angular2/src/core/compiler/element_injector.js]: done

File: modules/angular2/src/core/compiler/element_injector.js (hunk 3 above)
Comment (yjbanov) [inline modules/angular2/src/core/compiler/element_injector.js]: @tbosch actually, looks like this is not possible. The following should be legal:

```dart
@Component(
  selector: 'parent',
  publishAs: [Service]
)
class Parent extends Service {
  Parent(Service parentService) { ... }
}
```

Here the parent injects a `Service` from above, but publishes itself as a different `Service` implementation to its children. Therefore we cannot put the binding into the same injector where `injectables` go. We have to create a child injector.

Comment (tbosch) [inline modules/angular2/src/core/compiler/element_injector.js]: Mmh, true.
Is the following working before your change?

class Parent {
  Parent(Parent parent) { ... }
}

If yes, then I agree that we need a child injector. If no, then this was
not working before, i.e. does not seem to be a supported use case at the
moment....

Comment (yjbanov) [inline modules/angular2/src/core/compiler/element_injector.js]: Which is why I am adding the support, although not your example, but the example above. I added a test.

File: modules/angular2/test/core/compiler/element_injector_spec.js (hunk 4 above)
Comment (yjbanov) [inline modules/angular2/test/core/compiler/element_injector_spec.js]: Decided that this should not be supported by design. Added test to verify that.

**Pull request merged**